File: jclasslib/src/org/gjt/jclasslib/browser/detail/FixedListDetailPane.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail;

import org.gjt.jclasslib.browser.*;
import org.gjt.jclasslib.util.*;

import javax.swing.*;
import javax.swing.tree.*;
import java.awt.*;
import java.util.*;

/**
    Base class for all detail panes with a structure of
    a fixed number of key-value pairs arranged in a list.
    
    @author <a href="mailto:jclasslib@gmx.net">Ingo Kegel</a>
    @version $Revision: 1.2 $ $Date: 2001-05-31 13:15:25 $
*/
public abstract class FixedListDetailPane extends AbstractDetailPane {
    
    private java.util.List resizeListeners;
        
    // Visual components

    private java.util.List detailPaneEntries;

    public FixedListDetailPane(BrowserServices services) {
        super(services);
    }

    /**
        Add a detail entry consisting of a key value pair.
        @param key the key label
        @param value the value label
     */
    protected void addDetailPaneEntry(ExtendedJLabel key, ExtendedJLabel value) {
        addDetailPaneEntry(key, value, null);
    }

    /**
        Add a detail entry consisting of a key value pair with a associated comment
        which is made scrollable.
        @param key the key label
        @param value the value label
        @param comment the comment
     */
    protected void addDetailPaneEntry(ExtendedJLabel key,
                                      ExtendedJLabel value,
                                      ExtendedJLabel comment) {
                                          
        if (detailPaneEntries == null) {
            detailPaneEntries = new ArrayList();
        }
        
        detailPaneEntries.add(
                new DetailPaneEntry(key, value, comment)
            );
    }
    
    protected void setupComponent() {
        
        setupLabels();
        
        setLayout(new GridBagLayout());
        
        GridBagConstraints gKey = new GridBagConstraints();
        gKey.anchor = GridBagConstraints.NORTHWEST;
        gKey.insets = new Insets(1,10,0,10);
        
        GridBagConstraints gValue = new GridBagConstraints();
        gValue.gridx = 1;
        gValue.anchor = GridBagConstraints.NORTHEAST;
        gValue.insets = new Insets(1,0,0,5);

        GridBagConstraints gComment = new GridBagConstraints();
        gComment.gridx = 2;
        gComment.anchor = GridBagConstraints.NORTHWEST;
        gComment.insets = new Insets(1,0,0,5);
        gComment.fill = GridBagConstraints.HORIZONTAL;
        
        GridBagConstraints gCommentOnly = (GridBagConstraints)gComment.clone();
        gCommentOnly.gridx = 1;
        gCommentOnly.gridwidth = 2;
        
        GridBagConstraints gRemainder = new GridBagConstraints();
        gRemainder.gridx = 2;
        gRemainder.weightx = gRemainder.weighty = 1;
        gRemainder.fill = GridBagConstraints.BOTH;

        resizeListeners = new ArrayList();
        Iterator it = detailPaneEntries.iterator();
        while (it.hasNext()) {
            DetailPaneEntry entry = (DetailPaneEntry)it.next();
            if (entry == null) {
                continue;
            }

            gComment.gridy = gValue.gridy = ++gKey.gridy;
            if (entry.key != null) {
                add(entry.key, gKey);
            }
            if (entry.value != null) {
                add(entry.value, gValue);
            }
            if (entry.comment != null) {
                ScrollableJLabel scollableLabel = new ScrollableJLabel(entry.comment);
                
                services.addMaximizedListener(scollableLabel);
                resizeListeners.add(scollableLabel); 
                    
                add(scollableLabel, (entry.value == null) ? gCommentOnly : gComment);

                entry.comment.setAutoTooltip(true);
            }
            
        }
        
        gRemainder.gridy = gKey.gridy + 1;
        add(new JPanel(), gRemainder);
        
    }

    public void show(TreePath treePath) {
        for (int i = 0; i < resizeListeners.size(); i++) {
            ((ScrollableJLabel)resizeListeners.get(i)).resize(true);
        }
    }

    /**
        Setup all label and fill the <tt>detailPaneEntries</tt> list so that
        <tt>setupComponent</tt> can layout the pane.
     */
    protected abstract void setupLabels();
    
    
    private static class DetailPaneEntry {
        public final ExtendedJLabel key;
        public final ExtendedJLabel value;
        public final ExtendedJLabel comment;
        
        public DetailPaneEntry(ExtendedJLabel key,
                               ExtendedJLabel value,
                               ExtendedJLabel comment) {
            this.key = key;
            this.value = value;
            this.comment = comment;
        }
    }

    
}

File: modules/browser/src/main/java/org/gjt/jclasslib/browser/detail/FixedListDetailPane.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail

import org.gjt.jclasslib.browser.AbstractDetailPane
import org.gjt.jclasslib.browser.BrowserServices
import org.gjt.jclasslib.structures.Constant
import org.gjt.jclasslib.structures.attributes.BootstrapMethodsAttribute
import org.gjt.jclasslib.util.ExtendedJLabel
import org.gjt.jclasslib.util.GUIHelper
import org.gjt.jclasslib.util.TextDisplay

import javax.swing.*
import javax.swing.tree.TreePath
import java.awt.*
import java.util.ArrayList

abstract class FixedListDetailPane<T : Any>(private val elementClass: Class<T>, services: BrowserServices) : AbstractDetailPane(services) {

    private val scrollPane = JScrollPane(this).apply {
        GUIHelper.setDefaultScrollBarUnits(this)
        border = null
    }
    private var currentY = 0

    public override val wrapper: JComponent
        get() = scrollPane

    @JvmOverloads open protected fun addDetailPaneEntry(key: ExtendedJLabel, value: TextDisplay?, comment: TextDisplay? = null) {
        add(key, gc() {
            insets = Insets(1, 10, 0, 10)
        })
        if (value != null) {
            add(value as JComponent, gc() {
                gridx = 1
                insets = Insets(1, 0, 0, 5)
                if (comment == null) {
                    gridwidth = 2
                }
            })
        }
        if (comment != null) {
            add(comment as JComponent, gc() {
                if (value == null) {
                    gridx = 1
                    gridwidth = 2
                } else {
                    gridx = 2
                }
                insets = Insets(1, 0, 0, 5)
                fill = GridBagConstraints.HORIZONTAL
            })
            if (comment is ExtendedJLabel) {
                comment.autoTooltip = true
            }
        }
        nextLine()
    }

    protected fun nextLine() {
        currentY++
    }

    override fun setupComponent() {
        layout = GridBagLayout()
        addLabels()
        add(JPanel(), gc() {
            gridx = 2
            weightx = 1.0
            weighty = 1.0
            fill = GridBagConstraints.BOTH
        })
    }

    protected fun gc(config: GridBagConstraints.() -> Unit) = GridBagConstraints().apply {
        anchor = GridBagConstraints.NORTHWEST
        gridy = currentY
        config()
    }

    override fun show(treePath: TreePath) {
        scrollPane.viewport.viewPosition = Point(0, 0)
        element = getElement(treePath)
        element?.let { element -> showHandlers.forEach { it.invoke(element) } }
    }

    protected abstract fun addLabels()

    protected open fun addSpecial(gridy: Int): Int {
        return 0
    }

    protected val showHandlers = ArrayList<(element: T) -> Unit>()
    protected var element: T? = null

    override fun getElement(treePath: TreePath): T {
        return elementClass.cast(super.getElement(treePath))
    }

    protected fun addConstantPoolLink(name: String, indexResolver: (element: T) -> Int) {
        val nameLabel = linkLabel()
        val nameVerboseLabel = highlightLabel()
        addDetailPaneEntry(normalLabel(name), nameLabel, nameVerboseLabel)
        showHandlers.add { element ->
            constantPoolHyperlink(nameLabel, nameVerboseLabel, indexResolver(element))
        }
    }

    protected fun addAttributeLink(name: String, attributeClass: Class<BootstrapMethodsAttribute>, prefix: String, indexResolver: (element: T) -> Int) {
        val nameLabel = linkLabel()
        addDetailPaneEntry(normalLabel(name), nameLabel)
        showHandlers.add { element ->
            classAttributeIndexHyperlink(nameLabel, null, indexResolver(element), attributeClass, prefix)
        }
    }

    protected fun addCompositeDetail(name: String, textResolver: (element: T) -> Pair<String, String>) {
        val nameLabel = linkLabel()
        val nameVerboseLabel = highlightLabel()
        addDetailPaneEntry(normalLabel(name), nameLabel, nameVerboseLabel)
        showHandlers.add { element ->
            val texts = textResolver(element)
            nameLabel.text = texts.first
            nameVerboseLabel.text = texts.second
        }
    }

    protected fun addDetail(name: String, textResolver: (element: T) -> String) {
        val nameLabel = highlightLabel()
        addDetailPaneEntry(normalLabel(name), nameLabel)
        showHandlers.add { element ->
            nameLabel.text = textResolver(element)
        }
    }

    protected fun addMultiLineDetail(name: String, textResolver: (element: T) -> String) : LineControl<T> {
        val nameLabel = highlightTextArea()
        val keyLabel = normalLabel(name)
        addDetailPaneEntry(keyLabel, nameLabel)
        // TODO use multiline details and line control for all cases
        val lineControl = LineControl<T>(keyLabel, nameLabel)
        showHandlers.add { element ->
            nameLabel.text = textResolver(element)
            lineControl.show(element)
        }
        return lineControl
    }

    protected fun addClassElementOpener(constantResolver: (element: T) -> Constant) {
        if (services.canOpenClassFiles()) {
            val classElementOpener = ClassElementOpener(this)
            add(classElementOpener, gc() {
                weightx = 1.0
                anchor = GridBagConstraints.WEST
                insets = Insets(5, 10, 0, 10)
                gridx = 0
                gridwidth = 3
            })
            nextLine()
            showHandlers.add { element ->
                classElementOpener.setConstant(constantResolver(element))
            }
        }
    }
}
